### PR TITLE
Fix database inconsistent when admin change user email

### LINKED ISSUE #17519: Changing email account doesn't release that email
I set the admin account to xxx@mail.com, after built everything up, I change it to yyy@mail.com.
Now I try to set a new user with that email(xxx@mail.com), it always alerts the email is already existed.

### PR DESCRIPTION
**Pull request by lunny**
Fix #17519

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -1078,18 +1078,52 @@ func validateUser(u *User) error {
...
 
-func updateUser(e db.Engine, u *User) error {
+func updateUser(ctx context.Context, u *User, emailChanged bool) error {
 	if err := validateUser(u); err != nil {
 		return err
 	}
 
+	e := db.GetEngine(ctx)
+
+	if emailChanged {
```
Comment (zeripath) [inline models/user.go]: Why don't we just always do this?

Comment (lunny) [reply models/user.go]: It's unnecessary to always update if the primary email hasn't been changed. In fact, currently users cannot change them himself, only site admin could change them.

**Review by zeripath — APPROVED** [review]
This suffers from the select for update problems we have elsewhere but I guess it's the same so lgtm

File: models/user.go (hunk 1 above)
Comment (wxiaoguang) [reply models/user.go]: Then the variable name should be `changePrimaryEmail`

File: routers/api/v1/admin/user.go [hunk 2]
```diff
@@ -200,7 +201,9 @@ func EditUser(ctx *context.APIContext) {
 	if form.FullName != nil {
 		u.FullName = *form.FullName
 	}
+	var emailChanged bool
 	if form.Email != nil {
+		emailChanged = !strings.EqualFold(u.Email, *form.Email)
```
Comment (wxiaoguang) [inline routers/api/v1/admin/user.go]: This and below `strings.EqualFold` don't seem safe for unicode strings.

```
func main() {
 s := string([]byte{0xE2, 0x84, 0xAA})
 log.Print(s)
 log.Print("K" == s)
 log.Print(strings.EqualFold("K", s))
}
2021/11/09 18:27:42 K
2021/11/09 18:27:42 false
2021/11/09 18:27:42 true
```

Maybe we need something like ` util.StringEqualIncase`

**Comment by lunny** [comment]
depend on #17688

**Comment by lafriks** [comment]
Please resolve conflicts

File: routers/api/v1/admin/user.go [hunk 3]
```diff
@@ -202,12 +203,21 @@ func EditUser(ctx *context.APIContext) {
...
 			ctx.Error(http.StatusUnprocessableEntity, "", fmt.Errorf("email is not allowed to be empty string"))
 			return
 		}
+
+		if err := user_model.ValidateEmail(email); err != nil {
+			ctx.InternalServerError(err)
+			return
+		}
+
+		emailChanged = !strings.EqualFold(u.Email, email)
```
Comment (lunny) [inline routers/api/v1/admin/user.go:220]: We assume `ValidateEmail` will remove all invisible characters.

File: routers/api/v1/admin/user.go (hunk 2 above)
Comment (lunny) [reply routers/api/v1/admin/user.go]: I added `ValidateEmail` here, so I think it's that function's responsibility to remove all these invisible characters.

File: routers/web/admin/users.go [hunk 4]
```diff
@@ -332,6 +332,7 @@ func EditUserPost(ctx *context.Context) {
 
 	u.LoginName = form.LoginName
 	u.FullName = form.FullName
+	emailChanged := !strings.EqualFold(u.Email, form.Email)
```
Comment (wxiaoguang) [inline routers/web/admin/users.go:342]: `form.Email` seems to be validated by go-chi's EmailPattern:
```
EmailPattern        = regexp.MustCompile(`\A[\w!#$%&'*+/=?^_`+"`"+`{|}~-]+(?:\.[\w!#$%&'*+/=?^_`+"`"+`{|}~-]+)*@(?:[\w](?:[\w-]*[\w])?\.)+[a-zA-Z0-9](?:[\w-]*[\w])?\z`)
```
Impossible to understand what it does.

I don't know whether it is correct, whether it produces the same result as `ValidateEmail`

Comment (lunny) [reply routers/web/admin/users.go:342]: Yeah, we can discuss that but not block this PR?

**Review by wxiaoguang — APPROVED** [review]

File: routers/web/admin/users.go (hunk 4 above)
Comment (wxiaoguang) [reply routers/web/admin/users.go:342]: It's better to also use `ValidateEmail` here to match the behavior of API `EditUser`. 

LGTM.

Comment (lunny) [reply routers/web/admin/users.go:342]: done.

**Pull request merged**